Cubical/Structures/FreeCommAlgebra.agda

```diff
@@ -0,0 +1,398 @@
+                                                         (imageOf0Works) ⟩
+             (inducedMap P) + 0a                      ≡⟨ +-rid _ ⟩
+             (inducedMap P) ∎
+      in eq i
+    inducedMap (Construction.+-rinv P i) =
+      let eq : (inducedMap P - inducedMap P) ≡ (inducedMap (const 0r))
+          eq = (inducedMap P - inducedMap P) ≡⟨ +-rinv _ ⟩
+               0a                            ≡⟨ sym imageOf0Works ⟩
+               (inducedMap (const 0r))∎
+      in eq i
+    inducedMap (Construction.+-comm P Q i) = +-comm (inducedMap P) (inducedMap Q) i
+    inducedMap (P ·c Q) = inducedMap P · inducedMap Q
+    inducedMap (Construction.·-assoc P Q S i) = ·-assoc (inducedMap P) (inducedMap Q) (inducedMap S) i
+    inducedMap (Construction.·-lid P i) =
+      let eq = inducedMap (const 1r) · inducedMap P ≡⟨ cong (λ u → u · inducedMap P) imageOf1Works ⟩
+               1a · inducedMap P                    ≡⟨ ·-lid (inducedMap P) ⟩
+               inducedMap P ∎
+      in eq i
+    inducedMap (Construction.·-comm P Q i) = ·-comm (inducedMap P) (inducedMap Q) i
+    inducedMap (Construction.ldist P Q S i) = ·-ldist-+ (inducedMap P) (inducedMap Q) (inducedMap S) i
+    inducedMap (Construction.+HomConst s t i) = ⋆-ldist s t 1a i
+    inducedMap (Construction.·HomConst s t i) =
+      let eq = (s ·r t) ⋆ 1a       ≡⟨ cong (λ u → u ⋆ 1a) (·r-comm _ _) ⟩
+               (t ·r s) ⋆ 1a       ≡⟨ ⋆-assoc t s 1a ⟩
+               t ⋆ (s ⋆ 1a)        ≡⟨ cong (λ u → t ⋆ u) (sym (·-rid _)) ⟩
+               t ⋆ ((s ⋆ 1a) · 1a) ≡⟨ ⋆-rassoc t (s ⋆ 1a) 1a ⟩
+               (s ⋆ 1a) · (t ⋆ 1a) ∎
+      in eq i
+    inducedMap (Construction.0-trunc P Q p q i j) =
+      isSetAlgebra (CommAlgebra→Algebra A) (inducedMap P) (inducedMap Q) (cong _ p) (cong _ q) i j
+
+    inducedHom : AlgebraHom (CommAlgebra→Algebra (R [ I ])) (CommAlgebra→Algebra A)
+    inducedHom = algebrahom
+                   inducedMap
+                   (λ x y → refl)
+                   (λ x y → refl)
+                   imageOf1Works
+                   λ r x → (r ⋆ 1a) · inducedMap x ≡⟨ ⋆-lassoc r 1a (inducedMap x) ⟩
+                           r ⋆ (1a · inducedMap x) ≡⟨ cong (λ u → r ⋆ u) (·-lid (inducedMap x)) ⟩
+                           r ⋆ inducedMap x ∎
+
+  module _ (A : CommAlgebra R) where
+    open CommAlgebra A
+    open AlgebraTheory (CommRing→Ring R) (CommAlgebra→Algebra A)
+    open Construction using (var; const) renaming (_+_ to _+c_; -_ to -c_; _·_ to _·c_)
+
+    Hom = AlgebraHom (CommAlgebra→Algebra (R [ I ])) (CommAlgebra→Algebra A)
+    open AlgebraHom
+
+    _$_ : Hom → ⟨ R [ I ] ⟩a → ⟨ A ⟩a
+    f $ x = AlgebraHom.f f x
+
+    evaluateAt : Hom
+                 → I → ⟨ A ⟩a
+    evaluateAt (algebrahom f _ _ _ _) x = f (var x)
+
+    mapRetrievable : ∀ (φ : I → ⟨ A ⟩a)
+                     → evaluateAt (inducedHom A φ) ≡ φ
+    mapRetrievable φ = refl
+
+    proveEq : ∀ {X : Type ℓ} (isSetX : isSet X) (f g : ⟨ R [ I ] ⟩a → X)
```

Comment:
Yikes! This definition was complicated... I wish we could automate the cases for the higher constructors as they're all following the same pattern. @Saizan any thoughts about this after the discussion yesterday?

Reply:
It is long but not really complicated. And I wrote it down in a verbose way, with types of all the boundaries-paths of all the squares I used. So most of the code is just there to make it readable (for me).

Reply:
Yeah, it's pretty clear what's going on. I recently learned that you don't have to give these cases in Arend and I just wonder if we can automate them in Agda somehow...

Reply:
It would be nice if we could do this with a tactic, so it can just take the appropriate isOfHLevel proof as argument.

Might need some extension to the reflection machinery though.

Reply:
Actually maybe one can even prove a general theorem for solving this kind of goals parameterized by the hlevel. Doing it for prop, set and groupoid would be a good start.